### PR TITLE
Axe PersistedState interface from GatewayMetaState

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today `GatewayMetaState` implements `PersistedState` but it's an error to use
it as a `PersistedState` before it's been started, or if the node is
master-ineligible. It also holds some fields that are meaningless on nodes that
do not persist their states. Finally, it takes responsibility for both loading
the original cluster state and some of the high-level logic for writing the
cluster state back to disk.

This commit addresses these concerns by introducing a more specific
`PersistedState` implementation for use on master-eligible nodes which is only
instantiated if and when it's appropriate. It also moves the fields and
high-level persistence logic into a new `IncrementalClusterStateWriter` with a
more appropriate lifecycle.

Follow-up to #46326 and #46532

### CODE REVIEW

**Comment by original-brownbear** [comment]
@DaveCTurner why are we starting tis change in `7.x`? (I take it from the labels it is supposed to go into `master` as well?)

**Comment by DaveCTurner** [comment]
It's a good question @original-brownbear. The issue is that `7.x` maintains support for Zen1 (solely for `Zen1IT`) but this support is gone in `master` making metadata persistence much simpler there. I am doing this work on `7.x` and then forward-porting it after review to make sure I don't mess anything up with the Zen1/Zen2 interop.

**Review by andrershov — APPROVED** [review]
I've left a couple of comments to add a couple of comments, otherwise looks good!

**Review by original-brownbear — APPROVED** [review]
LGTM too, just a random NIT. (Didn't get through 100% of this but I trust Andrey did and the prod. changes look fine as far as my understanding fo this code goes :))

File: server/src/main/java/org/elasticsearch/gateway/IncrementalClusterStateWriter.java [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            try {
+                metaStateService.writeManifestAndCleanup(reason, manifest);
+                commitCleanupActions.forEach(Runnable::run);
+                finished = true;
+            } catch (WriteStateException e) {
+                // if Manifest write results in dirty WriteStateException it's not safe to remove
+                // new metadata files, because if Manifest was actually written to disk and its deletion
+                // fails it will reference these new metadata files.
+                // In the future, we might decide to add more fine grained check to understand if after
+                // WriteStateException Manifest deletion has actually failed.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/IncrementalClusterStateWriter.java]: :grin: yep see 485895153baa0085286f249ff9cd7412633cd18d

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 2]
```diff
@@ -49,124 +48,98 @@
...
+                // between the IndicesStore deleting the index folder and the CoordinationState concurrently trying to write the updated
+                // metadata into it. We could probably solve this with careful synchronization, but in fact there is no need.  The persisted
+                // state on master-ineligible data nodes is mostly ignored - it's only there to support dangling index imports, which is
+                // inherently unsafe anyway. Thus we can safely delay metadata writes on master-ineligible data nodes until applying the
+                // cluster state, which is what this does:
+                clusterService.addLowPriorityApplier(new GatewayClusterApplier(incrementalClusterStateWriter));
             }
+            persistedState.set(new InMemoryPersistedState(manifestClusterStateTuple.v1().getCurrentTerm(), manifestClusterStateTuple.v2()));
+        } else {
+            persistedState.set(new GatewayPersistedState(incrementalClusterStateWriter));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Sure, I added more detail in 485895153baa0085286f249ff9cd7412633cd18d.

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 3]
```diff
@@ -49,124 +48,98 @@
...
- * When instance of this class is created, constructor ensures that this version is compatible with state stored on disk and performs
- * state upgrade if necessary. Also it checks that atomic move is supported on the filesystem level, because it's a must for metadata
- * store algorithm.
- * Please note that the state being loaded when constructing the instance of this class is NOT the state that will be used as a
- * {@link ClusterState#metaData()}. Instead when node is starting up, it calls {@link #getMetaData()} method and if this node is
- * elected as master, it requests metaData from other master eligible nodes. After that, master node performs re-conciliation on the
- * gathered results, re-creates {@link ClusterState} and broadcasts this state to other nodes in the cluster.
+ * Loads (and maybe upgrades) cluster metadata at startup, and persistently stores cluster metadata for future restarts.
+ *
+ * When started, ensures that this version is compatible with the state stored on disk, and perform a state upgrade if necessary.  Note that
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Yep, thanks. I left that double space in there so there would be room for the missing `s` :wink: - see 485895153baa0085286f249ff9cd7412633cd18d.

**Pull request merged**